### ISSUE TITLE
instable build becuase of  incomplete download

### ISSUE THREAD

**Issue by syheliel** [frida/frida-rust#46]
The following error will appear when you have an incomplete download of `devkit tar.gz`, please add some completeness check when downloading.
```
Caused by:
  process didn't exit successfully: `/home/xxx/git/LibAFL/target/release/build/frida-gum-sys-04cb55fc4f28c053/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=event_sink.c
  cargo:rerun-if-changed=event_sink.h
  cargo:rerun-if-changed=invocation_listener.c
  cargo:rerun-if-changed=invocation_listener.h
  cargo:rustc-link-search=/home/xxx/.cargo/registry/src/mirrors.ustc.edu.cn-12df342d903acd47/frida-gum-sys-0.3.4

  --- stderr
  thread 'main' panicked at 'cannot extract the devkit tar.gz: Custom { kind: UnexpectedEof, error: TarError { desc: "failed to unpack `/home/xxx/git/LibAFL/target/release/build/frida-gum-sys-c7ba850327226163/out/libfrida-gum.a`", io: Custom { kind: UnexpectedEof, error: TarError { desc: "failed to unpack `libfrida-gum.a` into `/home/xxx/git/LibAFL/target/release/build/frida-gum-sys-c7ba850327226163/out/libfrida-gum.a`", io: Custom { kind: UnexpectedEof, error: "premature eof" } } } } }', /home/xxx/.cargo/registry/src/mirrors.ustc.edu.cn-12df342d903acd47/frida-build-0.2.0/src/lib.rs:55:14
  stack backtrace:
     0:     0x558de358fead - std::backtrace_rs::backtrace::libunwind::trace::h8f9ba895f18b09d0
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
     1:     0x558de358fead - std::backtrace_rs::backtrace::trace_unsynchronized::hc6a14c0ffe629597
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
     2:     0x558de358fead - std::sys_common::backtrace::_print_fmt::h0bced23525216d79
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/sys_common/backtrace.rs:66:5
     3:     0x558de358fead - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::heb53bc5f6fd72de1
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/sys_common/backtrace.rs:45:22
     4:     0x558de35b686c - core::fmt::write::h9eead92cdcbde24c
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/core/src/fmt/mod.rs:1198:17
     5:     0x558de35895f1 - std::io::Write::write_fmt::h4b47f879edc26800
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/io/mod.rs:1672:15
     6:     0x558de35917b5 - std::sys_common::backtrace::_print::hdda4f6fb9c1bea83
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/sys_common/backtrace.rs:48:5
     7:     0x558de35917b5 - std::sys_common::backtrace::print::h1334fdcbaa6ee9eb
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/sys_common/backtrace.rs:35:9
     8:     0x558de35917b5 - std::panicking::default_hook::{{closure}}::h4527c1ef3d428396
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:295:22
     9:     0x558de35914d6 - std::panicking::default_hook::hf30182a640d582ac
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:314:9
    10:     0x558de3591d46 - std::panicking::rust_panic_with_hook::h068be8eb179d2d55
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:698:17
    11:     0x558de3591c37 - std::panicking::begin_panic_handler::{{closure}}::h85625522130dfde4
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:588:13
    12:     0x558de3590364 - std::sys_common::backtrace::__rust_end_short_backtrace::hff27c063627d1102
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/sys_common/backtrace.rs:138:18
    13:     0x558de3591969 - rust_begin_unwind
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:584:5
    14:     0x558de2bd0063 - core::panicking::panic_fmt::hd728e2e35c66173a
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/core/src/panicking.rs:142:14
    15:     0x558de2bd01b3 - core::result::unwrap_failed::h1ff2879aca4f2f36
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/core/src/result.rs:1805:5
    16:     0x558de2f2f7cb - core::result::Result<T,E>::expect::h787e3e62754604b1
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/core/src/result.rs:1055:23
    17:     0x558de2c45f46 - frida_build::download_and_use_devkit::hf4218296eb375a8a
                                 at /home/xxx/.cargo/registry/src/mirrors.ustc.edu.cn-12df342d903acd47/frida-build-0.2.0/src/lib.rs:53:9
    18:     0x558de2bd09a0 - build_script_build::main::h032a7d6807c3bb79
                                 at /home/xxx/.cargo/registry/src/mirrors.ustc.edu.cn-12df342d903acd47/frida-gum-sys-0.3.4/build.rs:33:9
    19:     0x558de2bd11db - core::ops::function::FnOnce::call_once::ha296e293bdc10bf9
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/core/src/ops/function.rs:248:5
    20:     0x558de2bd16ce - std::sys_common::backtrace::__rust_begin_short_backtrace::h9e18af1ce3097940
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/sys_common/backtrace.rs:122:18
    21:     0x558de2bd14b1 - std::rt::lang_start::{{closure}}::ha77f430a4b7103ad
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/rt.rs:145:18
    22:     0x558de35840ee - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hcfe9c7d52c619978
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/core/src/ops/function.rs:280:13
    23:     0x558de35840ee - std::panicking::try::do_call::h8c4b802eb6e4184d
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:492:40
    24:     0x558de35840ee - std::panicking::try::h21616baf87e7165e
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:456:19
    25:     0x558de35840ee - std::panic::catch_unwind::h9080cdb55f15d458
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panic.rs:137:14
    26:     0x558de35840ee - std::rt::lang_start_internal::{{closure}}::hafbde1270c26291b
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/rt.rs:128:48
    27:     0x558de35840ee - std::panicking::try::do_call::h9d136e710d805d87
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:492:40
    28:     0x558de35840ee - std::panicking::try::h98e01b141a9fc81b
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panicking.rs:456:19
    29:     0x558de35840ee - std::panic::catch_unwind::h3ce8852ba8102f96
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/panic.rs:137:14
    30:     0x558de35840ee - std::rt::lang_start_internal::hf9507c4da6858514
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/rt.rs:128:20
    31:     0x558de2bd1480 - std::rt::lang_start::h04414e23e1cca25e
                                 at /rustc/7425fb293f510a6f138e82a963a3bc599a5b9e1c/library/std/src/rt.rs:144:17
    32:     0x558de2bd111c - main
    33:     0x7f37294fbd90 - __libc_start_call_main
                                 at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    34:     0x7f37294fbe40 - __libc_start_main_impl
                                 at ./csu/../csu/libc-start.c:392:3
    35:     0x558de2bd03b5 - _start
    36:                0x0 - <unknown>
    ```

**Comment by meme** [comment]
It's not clear to me how to reproduce. The download is being interrupted and there is no error? Or the download was interrupted and cached, and then a subsequent `cargo build` caused this issue?

**Comment by domenukk** [comment]
We could just delete the file and restart `download_and_use_devkit` if this line fails:
https://github.com/frida/frida-rust/blob/59cdd8a9f8be8f963028880b26c2cfbc6813eee8/frida-build/src/lib.rs#L55

**Comment by meme** [comment]
That sounds like a reasonable solution. Please send a patch if possible, as it's not clear how to repro this.

**Comment by meme** [comment]
This has been fixed! 4602be96a944a99ca8e5a8c7f9cea6a9ac531c37
